### PR TITLE
executor, planner, util: optimize query statements summary table by conditional pushdown

### LINKED ISSUE #26379: optimize query statements summary table by avoid decode plan by conditional pushdown
## Background

PR https://github.com/pingcap/tidb/pull/25843 enlarge the `tidb_stmt_summary_max_stmt_count` default value to 3000. It's great for diagnose problem since the `statements_summary ` can record more SQL. But it will also slow down the performance when query `statements_summary`. Mostly the reasons affecting performance is cause by `DecodePlan` for each row in `statements_summary`.

https://github.com/pingcap/tidb/pull/25892 already has optimized the performance by prune column. But it doesn't optimize for the following SQL which will be used in [tidb-dashboard](https://github.com/pingcap/tidb-dashboard): 

```sql
select * 
from `STATEMENTS_SUMMARY` 
where 
    digest='8019af26debae8aa7642c501dbc43212417b3fb14e6aec779f709976b7e521be' 
    and 
    unix_timestamp(summary_begin_time) = 1626769800;
```

The upper SQL need query all columns value, so https://github.com/pingcap/tidb/pull/25892 which use prune column method can't optimize it.

But we can use `conditional pushdown` to do optimization. As you can see the condition is `digest='8019af26debae8aa7642c501dbc43212417b3fb14e6aec779f709976b7e521be'  and   unix_timestamp(summary_begin_time) = 1626769800`. Mostly There is very little data that meets this condition, so we can use `conditional pushdown` to only fetch the row which meets the condition.

## Reference

How to do conditional pushdown for query system table?

- You can use `MemTablePredicateExtractor` to implements conditional pushdown.

You can reference the following PR which also do conditional pushdown for query system table:

- https://github.com/pingcap/tidb/pull/18092
- https://github.com/pingcap/tidb/pull/14840

##  mentor

-  [crazycs520](https://github.com/crazycs520)

If you have any questions about this issue, you can comment on this issue, I will reply you as soon as possible.

### PR DESCRIPTION
**Pull request by TszKitLo40**
Signed-off-by: Zijie Lu <wslzj40@gmail.com>

### What problem does this PR solve?

Issue Number: close #26379 

Bench:
Generate 3000 statements to filling `statemens summary` , and each SQL length is `10240 byte`.

1. Use [loadgen](https://github.com/crazycs520/loadgen) to generated load:
`bin/loadgen payload  gen-stmt --stmt-cnt 3000 --query-len 10240 --host 172.16.5.40 --port 4019  --concurrency 10`
  The generated SQL is look like this:
  `select * from t_gen_stmt_1 where a = 1 or b = 1 or c = 1 
  union all
  select * from t_gen_stmt_2 where a = 2 or b = 2 or c = 2 
  union all
  ...
  ...
  ...`

2. Wait generated statements(SQL) fill in the statemens summary.

3. Bench the SQL: select * from information_schema.cluster_statements_summary where digest = 'xxx';

Bellow is the bench result compare with the master:

|                      | Master |             This PR                |
| ------------ | --- | ------------------------------- |
| Query duration |  2.96s | 0.01s |

Problem Summary:

### What is changed and how it works?

What's Changed:
optimize query statements summary table by avoid decode plan by conditional pushdown
How it Works:
By using conditional pushdown, filter the digest in logical optimization phase.
### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None.
```

### CODE REVIEW

**Review by crazycs520 — COMMENTED** [review]
Got panic when i query following query:

```sql
select digest from `STATEMENTS_SUMMARY` where digest='';
```

File: planner/core/memtable_predicate_extractor.go [hunk 1]
```diff
@@ -204,7 +204,7 @@ func (helper extractHelper) extractCol(
 		}
 		var colName string
 		var datums []types.Datum // the memory of datums should not be reused, they will be put into result.
-		switch fn.FuncName.L {
+		switch helper.getStringFunctionName(fn) {
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go:207]: should also change the 'extractColOrExpr' function?

File: planner/core/pb_to_plan.go [hunk 2]
```diff
@@ -119,6 +120,8 @@ func (b *PBPlanBuilder) pbToTableScan(e *tipb.Executor) (PhysicalPlan, error) {
 			}
 		}
 		p.Extractor = extractor
+	case infoschema.ClusterTableStatementsSummary, infoschema.ClusterTableStatementsSummaryEvicted, infoschema.ClusterTableStatementsSummaryHistory:
```
Comment (crazycs520) [inline planner/core/pb_to_plan.go:123]:
```suggestion
	case infoschema.ClusterTableStatementsSummary, infoschema.ClusterTableStatementsSummaryHistory:
```

**Comment by crazycs520** [comment]
@djshow832  @jyz0309 PTAL

**Comment by crazycs520** [comment]
@TszKitLo40 Please add some unit tests.

File: planner/core/memtable_predicate_extractor.go [hunk 3]
```diff
@@ -1175,3 +1196,47 @@ func (e *TiFlashSystemTableExtractor) explainInfo(p *PhysicalMemTable) string {
...
+	}
+	return remained
+}
+
+func (e *StatementsSummaryExtractor) explainInfo(p *PhysicalMemTable) string {
+	if e.SkipRequest {
+		return "skip_request: true"
+	}
+	if !e.Enable {
+		return "All digests will be scanned"
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go:1239]: I think just return `""` will be ok here.

File: infoschema/tables_test.go [hunk 4]
```diff
@@ -1905,3 +1905,49 @@ func (s *testTableSuite) TestReferentialConstraints(c *C) {
...
+	c.Assert(len(tk.MustQuery(fmt.Sprintf("select digest from information_schema.cluster_statements_summary where digest = '%s' and digest = '%s'", digest1.String(), digest2.String())).Rows()), Equals, 0)
+	fillStatementCache()
+	tk.MustQuery(fmt.Sprintf("select digest from information_schema.statements_summary where digest in ('%s', '%s', '%s', '%s')", digest1.String(), digest2.String(), digest3.String(), digest4.String())).Sort().Check(testkit.Rows(digest1.String(), digest4.String(), digest2.String(), digest3.String()))
+	tk.MustQuery(fmt.Sprintf("select digest from information_schema.cluster_statements_summary where digest in ('%s', '%s', '%s', '%s')", digest1.String(), digest2.String(), digest3.String(), digest4.String())).Sort().Check(testkit.Rows(digest1.String(), digest4.String(), digest2.String(), digest3.String()))
+	fillStatementCache()
+	tk.MustQuery("select count(*) from information_schema.statements_summary where digest=''").Check(testkit.Rows("0"))
+	tk.MustQuery("select count(*) from information_schema.cluster_statements_summary where digest=''").Check(testkit.Rows("0"))
```
Comment (crazycs520) [inline infoschema/tables_test.go:1952]:
```suggestion
	tk.MustQuery("select count(*) from information_schema.statements_summary where digest=''").Check(testkit.Rows("0"))
	tk.MustQuery("select count(*) from information_schema.statements_summary where digest is null").Check(testkit.Rows("1"))
	tk.MustQuery("select count(*) from information_schema.cluster_statements_summary where digest=''").Check(testkit.Rows("0"))
	tk.MustQuery("select count(*) from information_schema.cluster_statements_summary where digest is null").Check(testkit.Rows("1"))
```

**Comment by crazycs520** [comment]
@qw4990 @djshow832 PTAL

**Review by djshow832 — COMMENTED** [review]
Please modify the PR title: it doesn't avoid decoding plans.

File: planner/core/rule_column_pruning.go [hunk 5]
```diff
@@ -304,6 +304,7 @@ func (p *LogicalMemTable) PruneColumns(parentUsedCols []*expression.Column) erro
 	for i := len(used) - 1; i >= 0; i-- {
 		if !used[i] && p.schema.Len() > 1 {
 			p.schema.Columns = append(p.schema.Columns[:i], p.schema.Columns[i+1:]...)
+			p.names = append(p.names[:i], p.names[i+1:]...)
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply planner/core/rule_column_pruning.go:307]: I think yes, isn't it? @crazycs520

File: util/stmtsummary/reader.go [hunk 6]
```diff
@@ -78,13 +80,19 @@ func (ssr *stmtSummaryReader) GetStmtSummaryCurrentRows() [][]types.Datum {
...
+			continue
+		}
+		record := ssr.getStmtByDigestRow(ssbd, beginTime)
 		if record != nil {
 			rows = append(rows, record)
 		}
 	}
-	if otherDatum := ssr.getStmtEvictedOtherRow(other); otherDatum != nil {
-		rows = append(rows, otherDatum)
+	if ssr.checker == nil {
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply util/stmtsummary/reader.go:92]: It has been covered in `infoschema/tables_test.go` L1952 and L1954

File: planner/core/memtable_predicate_extractor.go [hunk 7]
```diff
@@ -1175,3 +1196,47 @@ func (e *TiFlashSystemTableExtractor) explainInfo(p *PhysicalMemTable) string {
...
+		return nil
+	}
+	if digests.Count() > 0 {
+		e.Enable = true
+		e.Digests = digests
+	}
+	return remained
+}
+
+func (e *StatementsSummaryExtractor) explainInfo(p *PhysicalMemTable) string {
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply planner/core/memtable_predicate_extractor.go:1234]: done

File: planner/core/rule_column_pruning.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/rule_column_pruning.go:307]: Yes, it was a bug fix.

File: infoschema/tables_test.go [hunk 8]
```diff
@@ -1905,3 +1905,54 @@ func (s *testTableSuite) TestReferentialConstraints(c *C) {
...
+	tk.MustQuery(fmt.Sprintf("select digest from information_schema.cluster_statements_summary where digest in ('%s', '%s', '%s', '%s')", digest1.String(), digest2.String(), digest3.String(), digest4.String())).Sort().Check(testkit.Rows(digest1.String(), digest4.String(), digest2.String(), digest3.String()))
+	fillStatementCache()
+	tk.MustQuery("select count(*) from information_schema.statements_summary where digest=''").Check(testkit.Rows("0"))
+	tk.MustQuery("select count(*) from information_schema.statements_summary where digest is null").Check(testkit.Rows("1"))
+	tk.MustQuery("select count(*) from information_schema.cluster_statements_summary where digest=''").Check(testkit.Rows("0"))
+	tk.MustQuery("select count(*) from information_schema.cluster_statements_summary where digest is null").Check(testkit.Rows("1"))
+
+	// Test explain
+	tk.MustQuery("explain select digest from information_schema.statements_summary where digest is null").Check(testkit.Rows(`Selection_5 8000.00 root  isnull(Column#5)`, `└─MemTableScan_6 10000.00 root table:STATEMENTS_SUMMARY `))
```
Comment (crazycs520) [inline infoschema/tables_test.go:1957]: You can add following test in `executor/explainfor_test.go`.
```go
func (s *testSuite) TestExplainStatementsSummary(c *C) {
	tk := testkit.NewTestKitWithInit(c, s.store)
	tk.MustQuery("desc select * from information_schema.statements_summary").Check(testkit.Rows(
		`MemTableScan_4 10000.00 root table:STATEMENTS_SUMMARY `))
	tk.MustQuery("desc select * from information_schema.statements_summary where digest is null").Check(testutil.RowsWithSep("|",
		`Selection_5|8000.00|root| isnull(Column#5)`, `└─MemTableScan_6|10000.00|root|table:STATEMENTS_SUMMARY|`))
	tk.MustQuery("desc select * from information_schema.statements_summary where digest = 'abcdefg'").Check(testutil.RowsWithSep(" ",
		`MemTableScan_5 10000.00 root table:STATEMENTS_SUMMARY digests: ["abcdefg"]`))
	tk.MustQuery("desc select * from information_schema.statements_summary where digest in ('a','b','c')").Check(testutil.RowsWithSep(" ",
		`MemTableScan_5 10000.00 root table:STATEMENTS_SUMMARY digests: ["a","b","c"]`))
}
```

Comment (TszKitLo40) [reply infoschema/tables_test.go]: done

**Review by crazycs520 — APPROVED** [review]

**Comment by crazycs520** [comment]
@djshow832 PTAL

**Review by djshow832 — APPROVED** [review]

**Comment by djshow832** [comment]
Please resolve conflicts @TszKitLo40

**Pull request merged**